core/store/src/trie/mem/lookup.rs

```diff
@@ -1,34 +1,39 @@
+use std::sync::Arc;
+
+use super::metrics::MEM_TRIE_NUM_LOOKUPS;
 use super::node::{MemTrieNodePtr, MemTrieNodeView};
 use crate::NibbleSlice;
 use near_primitives::hash::CryptoHash;
 use near_primitives::state::FlatStateValue;
-use near_vm_runner::logic::TrieNodesCount;
-use std::collections::HashSet;
 
 /// Performs a lookup in an in-memory trie, while taking care of cache
 /// accounting for gas calculation purposes.
+///
+/// If `nodes_accessed` is provided, each trie node along the lookup path
+/// will be added to the vector as (node hash, serialized `RawTrieNodeWithSize`).
+/// Even if the key is not found, the nodes that were accessed to make that
+/// determination will be added to the vector.
+///
+/// The callback is used for:
+///  - Gas accounting of touched nodes
+///  - Populating the trie accounting cache (for legacy compatibility)
+///  - Recording accessed nodes, for state witness production.
 pub fn memtrie_lookup(
     root: MemTrieNodePtr<'_>,
     key: &[u8],
-    mut accessed_cache: Option<&mut HashSet<CryptoHash>>,
-    nodes_count: &mut TrieNodesCount,
+    mut nodes_accessed: Option<&mut Vec<(CryptoHash, Arc<[u8]>)>>,
```

Comment:
Great question. When a struct contains a &mut field, that field cannot be used to mutate if the struct itself is held with a non-mut &. This makes sense because otherwise, you can do this:

```
struct Mutator {
  target: &mut String,
}

let mut string = String::new();
let mutator = Mutator { target: string };
let m1 = &mutator;
let m2 = &mutator;
*m1.target = "abc".to_owned();
*m2.target = "def".to_owned();  // BAD: Two references can both mutate the string!
```

So yeah, to effectively use an Option<&mut T>, you always need the option itself to be mut.